currencies/rate_converter.go

```diff
@@ -0,0 +1,165 @@
+		ticker:                nil,
+		syncSourceURL:         syncSourceURL,
+		rates:                 nil,
+	}
+}
+
+// fetch allows to retrieve the currencies rates from the syncSourceURL provided
+func (rc *RateConverter) fetch() (*Rates, error) {
+	response, err := rc.httpClient.Get(rc.syncSourceURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	updatedRates := &Rates{}
+	err = json.NewDecoder(response.Body).Decode(&updatedRates)
+	if err != nil {
+		return nil, err
+	}
+	updatedRates.ComputeDataAsOf()
+
+	return updatedRates, err
+}
+
+// updateRates allows to update the internal currencies rates map with the one provided
+func (rc *RateConverter) updateRates(updatedRates *Rates) {
+	rc.Lock()
+	defer rc.Unlock()
+	rc.rates = updatedRates
+	now := time.Now()
+	rc.lastUpdated = &now
+}
+
+// Update updates the internal currencies rates from remote sources
+func (rc *RateConverter) Update() (*Rates, error) {
+	rates, err := rc.fetch()
+	if err == nil {
+		rc.Lock()
+		defer rc.Unlock()
+		rc.rates = rates
+		now := time.Now()
+		rc.lastUpdated = &now
+	}
+
+	return rates, err
+}
+
+// StartPeriodicFetching starts the periodic fetching at the given interval
+// triggers a first fetch when called before the first tick happen in order to initialize currencies rates map
+// returns a chan in which the number of data updates everytime a new update was done
+func (rc *RateConverter) StartPeriodicFetching() (chan int, error) {
+	rc.ticker = time.NewTicker(rc.fetchingInterval)
+	updatesTicksCount := 0
+	updatesNotifier := make(chan int)
+
+	go func() {
+		for {
+			select {
+			case <-rc.ticker.C:
+				go func() {
+					rates, err := rc.fetch()
+					if err == nil {
+						rc.updates <- rates
+					}
+					// TODO: Handle retry strategy
+				}()
+			case rates := <-rc.updates:
+				go func() {
+					rc.updateRates(rates)
+					updatesTicksCount++
+					updatesNotifier <- updatesTicksCount
+				}()
+			case <-rc.done:
+				if rc.ticker != nil {
+					rc.ticker.Stop()
+					rc.ticker = nil
+				}
+				return
+			default:
+			}
+		}
+	}()
+
+	// Trigger a first tick at the very begining
+	// the ticker will tick later
+	rates, err := rc.fetch()
+	if err != nil {
+		return nil, err
+	}
+	rc.updateRates(rates)
+
+	return updatesNotifier, nil
+}
+
+// StopPeriodicFetching stops the periodic fetching while keeping the latest currencies rates map
+func (rc *RateConverter) StopPeriodicFetching() {
+	rc.done <- true
+}
+
+// LastUpdated returns time when currencies rates were updated
+func (rc *RateConverter) LastUpdated() *time.Time {
+	return rc.lastUpdated
+}
+
+// Rates returns current rates
+func (rc *RateConverter) Rates() *Rates {
+	return rc.rates
+}
+
+// GetRate returns the conversion rate between two currencies
+// returns an error in case the conversion rate between the two given currencies is not in the currencies rates map
+func (rc *RateConverter) GetRate(from string, to string) (*float32, error) {
```

Comment:
Why a pointer return here?

Reply:
I wanted to be able to differenciate not found rate without having to pass a 0.
Maybe I can change the signature to remove the pointer:

`func (rc *RateConverter) GetRate(from string, to string) (float32, bool, error)`

Reply:
aha, I see.

Couldn't you use the `error` for that, though? If error is non-nil, the rate wasn't found. If the error is nil and rate is 0, then it's _really_ 0.

...in practice "real" 0 rates should be impossible, though.

Reply:
Yes I can, I was just thinking that errors and found / not found are not exactly the same in this context, but we can try.

Reply:
Done